content/blog/2021-07-05-LFX-Blog-Nupur.md

```diff
@@ -0,0 +1,131 @@
+application event log entry..."
+```
+
+gets translated to:
+
+```json
+{
+  "severity": "notice",
+  "facility": "local4",
+  "hostname": "mymachine.example.com",
+  "appname": "evntsog",
+  "msg": "BOMAn application event log entry...",
+  "procid": null,
+  "msgid": "ID47",
+  "protocol": "RFC5424",
+  "protocol_version": 1,
+  "structured_data": {
+              "exampleSDID@32473" :
+              [
+                {"iut": "3"},
+                {"eventSource": "Application"},
+                {"eventID": "1011"}
+              ]
+            },
+  "timestamp": 1616099400123000000
+}
+```
+
+Code for the syslog code can be found [here](https://github.com/tremor-rs/tremor-runtime/pull/856).
+
+### textual-length-prefix pre and postprocessor 
+
+In order to support syslog messages over TCP, it was needed to add support for the [RFC 5425](https://datatracker.ietf.org/doc/html/rfc5425) transport protocol, that contains a textual length prefix before each message.
+textual-length-prefix `preprocessor` and `postprocessor` were implemented to handle the buffers accordingly. The message starts with a number of digits, denoting the message length followed by a whitespace and then the message. The processor gets the length and then wait until the buffer is long enough, to extract the right amount of bytes.
+The implementation can be found [here](https://github.com/tremor-rs/tremor-runtime/pull/957).
+
+[Proptest](https://github.com/altsysrq/proptest) is something new and amazing I learnt while working on this. It is a property testing framework which allows to test certain properties of code for arbitrary inputs. We utilised this for testing the functioning of our preprocessor for all types of inputs possible.
+
+### TLS support for TCP
+
+Unlike UDP, Tremor did not support TLS over TCP onramp, which was needed to add. This work can be broadly divided into two parts:
+
+** Add support for receiving TLS encrypted data via TCP onramp **
+
+`tls` option was added to the tcp onramp configuration options which addresses the keys and certificate required for authentiction.
+
+An example of TCP onramp config with TLS is as follows:
+
+```yaml
+onramp:
+  - id: tls
+    type: tcp
+    preprocessors:
+      - lines
+    codec: string
+    config:
+      host: "127.0.0.1"
+      port: 65535
+      tls:
+        cert: "path/to/cert.pem"
+        key: "path/to/key.pem"
+```
+
+The code can be found [here](https://github.com/tremor-rs/tremor-runtime/pull/1055).
+
+** Add support for sending TLS encrypted data via TCP offramp **
+
+`tls` option added to offramp tls config contains either the tls config or boolean value indiacting the use of TLS session for transport level encryption. If false is provided then the default TCP stream will be used and if true is provided then TLS stream will be used with default certificates and domain same as hostname. Other option is to provide tls config with `domain` and `cafile`. In case of no domain, hostname will be used.
```

Comment:
```suggestion
`tls` option added to offramp tls config contains either the tls config or boolean value indiacting the use of TLS session for transport level encryption. If false is provided then the default TCP stream will be used and if true is provided then TLS stream will be used with default certificates and domain same as hostname. Other option is to provide tls config with `domain` and `cafile`. In case of `domain` not being specified, the hostname will be used.
```